### PR TITLE
[Foundation] Fix two Decimal members with respect to NaN

### PR DESCRIPTION
**Pull request by xwu**
This PR fixes `Decimal.magnitude` so that `Decimal.nan.magnitude` is no longer incorrectly `0`.

It also incorporates changes from apple/swift-corelibs-foundation#1759, which has the salient effect of correcting an error unique to this overlay where `(0 as Decimal).doubleValue` returned `.nan`.

Resolves [SR-10938](https://bugs.swift.org/browse/SR-10938).

### CODE REVIEW

**Comment by xwu** [comment]
@swift-ci Please test OS X platform

File: stdlib/public/Darwin/Foundation/Decimal.swift [hunk 1]
```diff
@@ -150,20 +150,31 @@ public func pow(_ x: Decimal, _ y: Int) -> Decimal {
...
-        d = d * 65536 + Double(_mantissa.6)
-        d = d * 65536 + Double(_mantissa.5)
-        d = d * 65536 + Double(_mantissa.4)
-        d = d * 65536 + Double(_mantissa.3)
-        d = d * 65536 + Double(_mantissa.2)
-        d = d * 65536 + Double(_mantissa.1)
-        d = d * 65536 + Double(_mantissa.0)
+
+        var d = 0.0
+        for idx in stride(from: min(_length, 8), to: 0, by: -1) {
```
Comment (stephentyrone) [inline stdlib/public/Darwin/Foundation/Decimal.swift]: `(0 ..< _length).reversed()`, use `idx` instead of `idx - 1` below?

**Comment by xwu** [comment]
@swift-ci Please test

**Comment by xwu** [comment]
@swift-ci Please test

**Comment by xwu** [comment]
(Reviewed by @spevans over at swift-corelibs-foundation. Will squash and merge.)

**Pull request merged**